msprime/demography.py

```diff
@@ -36,11 +37,309 @@
 logger = logging.getLogger(__name__)
 
 
+# TODO can we change this to be an attrs object instead?
+# Yes, we just need to make sure that we support tuples as input also
+# to preserve backwards compatability.
+Sample = collections.namedtuple("Sample", ["population", "time"])
+
+
+@attr.s
+class Demography:
+    """
+    A description of a demographic model for an msprime simulation.
+
+    TODO document properly.
+    """
+
+    populations = attr.ib(factory=list)
+    migration_matrix = attr.ib(default=None)
+    events = attr.ib(factory=list)
+
+    @property
+    def num_populations(self):
+        return len(self.populations)
+
+    def validate(self):
+        """
+        Checks the demography looks sensible and raises errors/warnings
+        appropriately.
+        """
+        migration_matrix = np.array(self.migration_matrix)
+        N = self.num_populations
+        if migration_matrix.shape != (N, N):
+            raise ValueError(
+                "migration matrix must be a N x N square matrix encoded "
+                "as a list-of-lists or numpy array, where N is the number "
+                "of populations. The diagonal "
+                "elements of this matrix must be zero. For example, a "
+                "valid matrix for a 3 population system is "
+                "[[0, 1, 1], [1, 0, 1], [1, 1, 0]]"
+            )
+
+        for event in self.events:
+            if not isinstance(event, DemographicEvent):
+                raise TypeError(
+                    "Demographic events must be a list of DemographicEvent "
+                    "instances sorted in non-decreasing order of time."
```

Comment:
It probably should check that events are sorted, but I don't think we'd want to do the sorting internally, since some demographic models will be sensitive to the order of events that occur at the same time, right?

Reply:
That should be fine, if a *stable* sorting algorithm is used. Looks like that's the default in python.
https://docs.python.org/3/howto/sorting.html#sort-stability-and-complex-sorts